pkg/logs/pipeline/pipeline.go

```diff
@@ -69,26 +66,69 @@ func NewPipeline(outputChan chan *message.Message, processingRules []*config.Pro
 	processor := processor.New(inputChan, senderChan, processingRules, encoder, diagnosticMessageReceiver)
 
 	return &Pipeline{
-		InputChan: inputChan,
-		processor: processor,
-		sender:    sender,
+		InputChan:    inputChan,
+		processor:    processor,
+		mainSender:   mainSender,
+		backupSender: backupSender,
 	}
 }
 
 // Start launches the pipeline
 func (p *Pipeline) Start() {
-	p.sender.Start()
+	p.mainSender.Start()
+	if p.backupSender != nil {
+		p.backupSender.Start()
+	}
 	p.processor.Start()
 }
 
 // Stop stops the pipeline
 func (p *Pipeline) Stop() {
 	p.processor.Stop()
-	p.sender.Stop()
+	p.mainSender.Stop()
+	if p.backupSender != nil {
+		p.backupSender.Stop()
+	}
 }
 
 // Flush flushes synchronously the processor and sender managed by this pipeline.
 func (p *Pipeline) Flush(ctx context.Context) {
-	p.processor.Flush(ctx) // flush messages in the processor into the sender
-	p.sender.Flush(ctx)    // flush the sender
+	p.processor.Flush(ctx)  // flush messages in the processor into the sender
+	p.mainSender.Flush(ctx) // flush the sender
+	if p.backupSender != nil {
+		p.backupSender.Flush(ctx)
+	}
+}
+
+func getMainDestinations(endpoints *config.Endpoints, destinationsContext *client.DestinationsContext) *client.Destinations {
+	if endpoints.UseHTTP {
+		main := http.NewDestination(endpoints.Main, http.JSONContentType, destinationsContext, endpoints.BatchMaxConcurrentSend)
+		additionals := []client.Destination{}
+		for _, endpoint := range endpoints.Additionals {
+			additionals = append(additionals, http.NewDestination(endpoint, http.JSONContentType, destinationsContext, endpoints.BatchMaxConcurrentSend))
+		}
+		return client.NewDestinations(main, additionals)
+	}
+	main := tcp.NewDestination(endpoints.Main, endpoints.UseProto, destinationsContext)
+	additionals := []client.Destination{}
+	for _, endpoint := range endpoints.Additionals {
+		additionals = append(additionals, tcp.NewDestination(endpoint, endpoints.UseProto, destinationsContext))
+	}
+	return client.NewDestinations(main, additionals)
+}
+
+func getBackupDestinations(endpoints *config.Endpoints, destinationsContext *client.DestinationsContext) *client.Destinations {
+	if endpoints.UseHTTP {
+		backup := http.NewDestination(*endpoints.Backup, http.JSONContentType, destinationsContext, endpoints.BatchMaxConcurrentSend)
+		return client.NewDestinations(backup, []client.Destination{})
+	}
+	backup := tcp.NewDestination(*endpoints.Backup, endpoints.UseProto, destinationsContext)
+	return client.NewDestinations(backup, []client.Destination{})
+}
```

Comment:
nit: I think these two functions havesome code duplication that could be extracted out.

Reply:
I'll think about it a bit more. but these are actually pretty hard to combine. There's just enough different stuff that im not sure it would be more readable (additionals vs no additionals for example). I mostly moved this logic here to make the pipeline code more readable - let me know if you had something else in mind.